graylog2-server/src/main/java/org/graylog2/bootstrap/preflight/PasswordSecretPreflightCheck.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog2.bootstrap.preflight;
+
+import jakarta.inject.Inject;
+import org.graylog2.plugin.cluster.ClusterConfigService;
+import org.graylog2.security.encryption.EncryptedValue;
+import org.graylog2.security.encryption.EncryptedValueService;
+
+import javax.annotation.Nonnull;
+import java.util.Optional;
+
+/**
+ * This preflight check exists to validate that every node has the same password_secret value configured.
+ * The first node in the cluster will persist a known and encoded secret in the cluster_config collection.
+ * Every other node is then testing if it's possible to read and decode the value. If not, then the password_secret
+ * is not matching and we'd get problems later during runtime. So it's better to stop the startup and report
+ * an explicit and readable error.
+ */
+public class PasswordSecretPreflightCheck implements PreflightCheck {
+
+
+    private static final String KNOWN_VALUE = "graylog";
+
+    private final ClusterConfigService clusterConfigService;
+
+    private final EncryptedValueService encryptionService;
+
+
+    @Inject
+    public PasswordSecretPreflightCheck(ClusterConfigService clusterConfigService, EncryptedValueService encryptionService) {
+        this.clusterConfigService = clusterConfigService;
+        this.encryptionService = encryptionService;
+    }
+
+    @Override
+    public void runCheck() throws PreflightCheckException {
+        final PreflightEncryptedSecret encryptedValue = clusterConfigService.get(PreflightEncryptedSecret.class);
+        Optional.ofNullable(encryptedValue)
+                .ifPresentOrElse(this::validateSecret, this::persistSecret);
+    }
+
+    private void persistSecret() {
+        final EncryptedValue encryptedValue = encryptionService.encrypt(KNOWN_VALUE);
+        clusterConfigService.write(new PreflightEncryptedSecret(encryptedValue));
+    }
+
+    private void validateSecret(@Nonnull PreflightEncryptedSecret preflightEncryptedSecret) {
+        final EncryptedValue encryptedSecret = preflightEncryptedSecret.encryptedSecret();
+
+        try {
+            final String decrypted = encryptionService.decrypt(encryptedSecret);
```

Comment:
unfortunately, our implementation in `AESTools` doesn't guarantee that the secret is identical as it modifies the key. You can reproduce that by removing a single letter from the end of a long password secret. This will still lead to problems with the keystore password.

Reply:
I don't think this is a problem related to this PR. For the mongodb entries encryption, this will detect any problem. Keystore reading will be a local problem that we can (and should) capture by proper error handling.

Reply:
I am not sure. I would expect the check to make sure that password_secrets are identical on all nodes, no matter if they are used for encryption or anything else. This is also what the class' javadoc implies.
With the given check, there is still a margin for error which is difficult to track down. Something that we wanted to avoid with this check.

Reply:
How could we fix that?

Reply:
Haven't thought it through completely yet, but what if we don't encrypt a constant value but the secret itself? Then, if if you can decrypt it with a shorter secret, you could still make sure that they are identical

Reply:
Implemented that :+1: